gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,197 @@
+  }
+
+  /**
+   * Returns AUTO, NEVER or ALWAYS. NEVER means always use regular endpoint; ALWAYS means always use
+   * mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists and should
+   * be used.
+   */
+  public UseMtlsEndpoint useMtlsEndpoint() {
+    String useMtlsEndpoint = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.NEVER;
+    } else if ("always".equals(useMtlsEndpoint)) {
+      return UseMtlsEndpoint.ALWAYS;
+    }
+    return UseMtlsEndpoint.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException ignored) {
+      // file doesn't exist
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException {
+    try {
+      Process process = processProvider.createProcess(metadata);
+
+      // Run the command and timeout after 1000 milliseconds.
+      int exitCode = runCertificateProviderCommand(process, 1000);
+      if (exitCode != 0) {
+        throw new IOException("Cert provider command failed with exit code: " + exitCode);
+      }
+
+      // Create mTLS key store with the input certificates from shell command.
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
+    }
+  }
+
+  @VisibleForTesting
+  static List<String> extractCertificateProviderCommand(InputStream contextAwareMetadata)
+      throws IOException {
+    JsonParser parser = new GsonFactory().createJsonParser(contextAwareMetadata);
+    ContextAwareMetadataJson json = parser.parse(ContextAwareMetadataJson.class);
+    return json.getCommands();
+  }
+
+  @VisibleForTesting
+  static int runCertificateProviderCommand(Process commandProcess, long timeoutMilliseconds)
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    long remainTime = timeoutMilliseconds;
+    boolean terminated = false;
+
+    while (remainTime > 0) {
+      Thread.sleep(Math.min(remainTime + 1, 100));
+      remainTime -= System.currentTimeMillis() - startTime;
+
+      try {
+        // Check if process is terminated by polling the exitValue, which throws
+        // IllegalThreadStateException if not terminated.
+        commandProcess.exitValue();
+        terminated = true;
+        break;
+      } catch (IllegalThreadStateException ignored) {
+      }
+    }
+
+    if (!terminated) {
```

Comment:
not sure you need the temrmnated variable; could perhaps pull it into the try block

Reply:
Yes the terminated variable is needed here. 
In the while loop, we keep calling `commandProcess.exitValue()` to check if the process is done. Every time this method throws `IllegalThreadStateException` exception if process is not done, so we need the try block inside the while loop. We exit the while loop if process is done or timeout. So this variable cannot be pulled into the try block as well.

Reply:
Move line 195, the return statement , to replace lines 183-185. Then you can remove the terminated variable and the if statement here.

Reply:
updated the code. Thanks!